ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestGeoSupport.java

```diff
@@ -0,0 +1,242 @@
+package denominator.ultradns;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.TreeSet;
+
+import javax.inject.Named;
+
+import dagger.Module;
+import dagger.Provides;
+import denominator.ultradns.model.DirectionalGroup;
+import denominator.ultradns.model.Region;
+import denominator.ultradns.util.RRSetUtil;
+import denominator.ultradns.util.RegionUtil;
+import org.apache.commons.lang.StringUtils;
+
+@Module(injects = UltraDNSRestGeoResourceRecordSetApi.Factory.class, complete = false)
+public class UltraDNSRestGeoSupport {
+
+  private UltraDNSRest api = null;
+
+  static final int MAX_EFFECTIVE_CODES_PER_API_CALL = 100;
+
+  public UltraDNSRestGeoSupport() { }
+
+  public UltraDNSRestGeoSupport(UltraDNSRest api) {
+    this.api = api;
+  }
+
+  @Provides
+  @Named("geo")
+  Map<String, Collection<String>> regions(UltraDNSRest api1) {
+    Map<String, Collection<String>> availableRegions = new TreeMap<String, Collection<String>>();
+
+    Collection<Region> topLevelRegions = buildRegionHierarchyByCallingUltraDNSRest(api1);
+    for (Region topLevelRegion : topLevelRegions) {
+      Map<String, Collection<String>> regionHierarchy = RegionUtil.getRegionHierarchy(topLevelRegion);
+      for (Map.Entry<String, Collection<String>> regionSubregions : regionHierarchy.entrySet()) {
+        availableRegions.put(regionSubregions.getKey(), regionSubregions.getValue());
+      }
+    }
+    return availableRegions;
+  }
+
+  Map<Region, Collection<Region>> regionsAsRegions() {
+    Map<Region, Collection<Region>> availableRegions = new TreeMap<Region, Collection<Region>>();
+
+    Collection<Region> topLevelRegions = buildRegionHierarchyByCallingUltraDNSRest(api);
+    for (Region topLevelRegion : topLevelRegions) {
+      Map<Region, Collection<Region>> regionHierarchy = RegionUtil.getRegionHierarchyAsRegions(topLevelRegion);
+      for (Map.Entry<Region, Collection<Region>> regionSubregions : regionHierarchy.entrySet()) {
+        availableRegions.put(regionSubregions.getKey(), regionSubregions.getValue());
+      }
+    }
+
+    return availableRegions;
+  }
+
+  private Collection<Region> buildRegionHierarchyByCallingUltraDNSRest(UltraDNSRest api2) {
+    Collection<Collection<Region>> response = api2.getAvailableRegions("");
+    Collection<Region> topLevelRegions = response.iterator().next();
+    TreeSet<String> sortedEffectiveCodes = getSortedEffectiveCodes(topLevelRegions);
+    String commaSeparatedCodes = getCommaSeparatedEffectiveCodes(sortedEffectiveCodes);
+
+    response = api2.getAvailableRegions(commaSeparatedCodes);
```

Comment:
Can the topLevelRegions be > 100. I checked the responses and they don't appear to be. Do you expect that they will be > 100 in the future?